Initial EQL rest API implementation

aleksmaus opened the pull request:
It's my first PR, need a bit of feedback on the code shape and direction.
Specifically had to poke through RBACEngine.java and “mark” the request with CompositeIndicesRequest interface, not sure if this is the right approach.

aleksmaus commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -212,6 +212,7 @@ private static boolean shouldAuthorizeIndexActionNameOnly(String action, Transpo
             case "indices:data/write/reindex":
             case "indices:data/read/sql":
             case "indices:data/read/sql/translate":
+            case "indices:data/read/eql":
```
was not sure if there is a better approach instead of this ^

imotov replied:
I am not sure we need this here. I don't see a scenario when we are going to resolve the index name based on the content of the request itself.

aleksmaus replied:
Removing this change, replacing with more suitable solution after discussion with Igor.

aleksmaus commented on server/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -887,7 +887,7 @@ public boolean equals(Object obj) {
                 && Objects.equals(primaryTerm, other.primaryTerm)
                 && Objects.equals(source, other.source)
-                && Objects.equals(fields, other.fields)
+                && Objects.equals(getFields(), other.getFields())
```
A possible fix for equals as a separate commit here, can remove if not appropriate or should it's own separate PR.
Sharing the same SearchHit structures in the eql resful API implementation, and the equality checks would fail without this  change in place.
Following the same thing that was done a line below with highlightFields accessors.

imotov replied:
It might be a bit cleaner to make this a separate PR to master and 7.x and then merge it in. In any case, I think this requires a modification to SearchHitTests that would trigger the issue that you are trying to fix here.

aleksmaus replied:
Sounds good, will update this PR once if goes through the master.

aleksmaus replied:
This file change is removed now from PR.

aleksmaus commented:
* Removed the SearchHit.java change. It will go through master branch.
* Updated the unit test for now to always create SearchHit with fields.
* Squashed and force pushed.

aleksmaus commented:
Pulled the latest master to feature/eql and rebased my PR against it, cause of build failures due to what  java version
`21:07:03 Caused by: org.gradle.api.GradleException: JAVA13_HOME required to run task:
21:07:03 task ':distribution:bwc:minor:buildBwcOssLinuxTar'`

master branch was on Java 13 already

elasticmachine commented:
Pinging @elastic/es-search (:Search/EQL)

imotov commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchRequest.java:
```diff
@@ -0,0 +1,242 @@
+
+    @Override
+    public ActionRequestValidationException validate() {
+        return null;
```
Since we rely on the fact that index, timestampField, eventTypeField, etc cannot be null and fetchSize cannot be negative, we should validate that it's indeed the case. We could also make serialization more forgiving and avoid serializing the same default values on every request.

aleksmaus replied:
Will add validation code

imotov commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/RestEqlSearchAction.java:
```diff
@@ -0,0 +1,60 @@
+        // Not sure yet if we will always have index in the path or not
+        // TODO: finalize the endpoints
+        controller.registerHandler(GET, "/{index}/_eql/search", this);
+        controller.registerHandler(GET, "/_eql/search", this);
```
In most cases, if the index name is not specified on the URL we search all indices or use index specified inside the request. In this case I don't think we will ever search all indices and I don't see any provisions from specifying the index name inside the request. So, I think we can remove this version.

aleksmaus replied:
Yeah, there was like going back-n-forth on this, so included both versions. Will remove the latter.

imotov commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:
```diff
@@ -49,15 +49,15 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
          * like everything else. We want to stick as closely as possible to
          * Elasticsearch's defaults though, while still layering in ways to
-         * control the output more easilly.
+         * control the output more easily.
```
I would move this file into a separate trivial PR.

aleksmaus replied:
Will revert this file change from this PR.

aleksmaus commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java:
```diff
@@ -3,11 +3,43 @@
+public class EqlPlugin extends Plugin implements ActionPlugin {
+    @Override
+    public List<ActionHandler<? extends ActionRequest, ? extends ActionResponse>> getActions() {
+        return Arrays.asList(
```
Thought there was a possibility of one more "translate" action down the road.

aleksmaus commented:
@elasticmachine run elasticsearch-ci/1
